qi-lib/flow/impl.rkt

```diff
@@ -210,26 +198,14 @@
   (λ args
     (apply values (zip-with call fs args))))
 
-(define (~all? . args)
-  (match args
-    ['() #t]
-    [(cons v vs)
-     (and v (apply all? vs))]))
-
-(define all? (compose not not ~all?))
-
-(define (~any? . args)
-  (match args
-    ['() #f]
-    [(cons v vs)
-     (or v (apply any? vs))]))
-
-(define any? (compose not not ~any?))
+(define (all? . args)
+  (and (for/and ([v (in-list args)]) v) #t))
 
-(define (~none? . args)
-  (not (apply any? args)))
+(define (any? . args)
+  (and (for/or ([v (in-list args)]) v) #t))
 
-(define none? (compose not not ~none?))
+(define (none? . args)
+  (not (for/or ([v (in-list args)]) v)))
```

Comment:
Nice, these do seem like they would be faster. Is the performance difference apparent using the profilers (e.g. `make profile-selected-forms`)?

Reply:
Here's how the two versions perform with a large number of arguments:

```racket
(require qi)

(displayln 'all?)
(time (~> (#t) (fanout 10000) all?))
(newline)

(displayln 'any?)
(time (~> (#f) (fanout 10000) any?))
(newline)

(displayln 'none?)
(time (~> (#f) (fanout 10000) none?))
(newline)
```

Original
```
all?
cpu time: 566 real time: 568 gc time: 48
#t

any?
cpu time: 564 real time: 566 gc time: 21
#f

none?
cpu time: 541 real time: 543 gc time: 18
#t
```

New
```
all?
cpu time: 0 real time: 0 gc time: 0
#t

any?
cpu time: 0 real time: 0 gc time: 0
#f

none?
cpu time: 0 real time: 0 gc time: 0
#t
```

Reply:
I think it might be because the`apply` is called n^2 times in the original version (n is the number of arguments)

Reply:
Why n^2? Isn't it n times?

Reply:
Yeah, I wrote it wrong, it seems that the complexity of apply is n, so the program's complexity is n^2 (n + n-1 + ... + 1). And if we use 10e5, it is much slower than 10e4.

Reply:
```racket
(require qi)

(displayln 'all?)
(time (~> (#t) (fanout 100000) all?))
(newline)
```

Original
```
all?
cpu time: 75784 real time: 76085 gc time: 12796
#t
```

New
```
all?
cpu time: 16 real time: 16 gc time: 8
#t
```

Reply:
Qi seems to use `apply` recursively in many places, and I'm a little concerned if there are similar problems elsewhere.

Reply:
Interesting, nice catch! And yeah there could easily be other cases affected by this.